### PR TITLE
Uplift internal linting

### PR DESCRIPTION
**Pull request by AaronMoat**
Replacing https://github.com/seek-oss/skuba/pull/1346, but same motivation:

> Currently, if .gitignore/other ignore files need to be fixed by format, you can get away with never having them fixed because lint never fails. I've seen cases where they're not pushed with autofix (unsure if always or just if no write perms?)
> ... 
> Fairly important to get right before pnpm / managing npmrc

Goal is:

- Lint is readonly, but autofixes still work in CI
- format and lint do the same things (but readonly vs write)
- Nice output 

I'm still newish to this codebase, so there may be crimes here, and my manual and automated testing is somewhat limited, so things to look out for tests to add (or run) would be appreciated.

### CODE REVIEW

**Review by 72636c — COMMENTED** [review]
Description makes sense 👍. Just opining on the TODOs, I'll take a more detailed look later.

File: src/cli/lint/upgrade/index.ts [hunk 1]
```diff
@@ -52,29 +53,58 @@ export const upgradeSkuba = async () => {
   const manifestVersion = (manifest.packageJson.skuba as SkubaPackageJson)
     .version;
 
-  // We are up to date, avoid apply patches
+  // We are up to date, skip patches
   if (gte(manifestVersion, currentVersion)) {
-    return;
+    return { ok: true, fixable: false };
```
[the remark this answers is not in the document]
Comment (AaronMoat) [reply src/cli/lint/upgrade/index.ts]: Ended up collecting annotations as we go so a string union won't work anymore

File: src/cli/lint/upgrade/index.ts [hunk 2]
```diff
@@ -52,29 +53,58 @@ export const upgradeSkuba = async () => {
...
+
+    logger.warn(
+      `skuba has patches to apply. Run ${logger.bold(
+        packageManager.exec,
+        'skuba',
+        'format',
+      )} to run them. ${logger.dim('skuba-patches')}`,
+    );
+
+    // TODO: Do we want to declare patches as optional?
```
[the remark this answers is not in the document]
Comment (AaronMoat) [reply src/cli/lint/upgrade/index.ts]: Yep. 

The other part of the story is that some people might _never_ run skuba format - using prettier/eslint in vscode to do the bulk of things - and so they won't get any patches unless CI fails.

File: src/cli/format.ts [hunk 3]
```diff
@@ -5,15 +5,16 @@ import { createLogger, log } from '../utils/logging';
...
-import { tryRefreshIgnoreFiles } from './configure/refreshIgnoreFiles';
-import { upgradeSkuba } from './configure/upgrade';
+import { internalLint } from './lint/internal';
 
 export const format = async (args = process.argv.slice(2)): Promise<void> => {
-  await Promise.all([tryRefreshIgnoreFiles(), upgradeSkuba()]);
+  log.plain(chalk.blueBright('skuba lints'));
+  const internal = await internalLint('format', { debug: false, serial: true });
 
   const debug = hasDebugFlag(args);
```
Comment (72636c) [inline src/cli/format.ts:14]:
```suggestion
  const debug = hasDebugFlag(args);

  log.plain(chalk.blueBright('skuba lints'));
  const internal = await internalLint('format', { debug, serial: true });

```

File: src/cli/configure/upgrade/patches/7.3.1/index.ts [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
-};
+export const patches: Patches = [
+  {
+    apply: tryAddEmptyExports,
+    description:
+      'Add empty exports to Jest files for compliance with TypeScript isolated modules',
+  },
+  {
+    apply: tryPatchRenovateConfig,
+    description: 'Update renovate config for private Renovate auth',
```
Comment (72636c) [inline src/cli/configure/upgrade/patches/7.3.1/index.ts:16]:
```suggestion
    description: 'Update Renovate config to support private SEEK packages',
```

File: src/cli/configure/upgrade/patches/7.3.1/index.ts [hunk 5]
```diff
@@ -1,14 +1,26 @@
...
+  },
+  {
+    apply: tryPatchDockerfile,
+    description: 'Swap node distroless Docker image to -debian11 variant',
```
Comment (72636c) [inline src/cli/configure/upgrade/patches/7.3.1/index.ts:20]:
```suggestion
    description: 'Upgrade Node.js Distroless Docker image to -debian11 variant',
```

File: .changeset/five-dancers-pump.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+---
+
+lint: Overhaul skuba's internal linting system
+
+Internal (skuba) linting is now promoted to a top-level tool alongside ESLint, tsc, and Prettier.
+
+This fixes issues where skuba would not fail a `lint` check but silently make changes.
+These changes may never end up being committed and causes noise when running `lint` or `format` later.
+
+Now, lints report whether changes need to be made and are applied in `format` or autofix modes (in CI).
```
Comment (72636c) [inline .changeset/five-dancers-pump.md:12]:
```suggestion
lint: Overhaul internal linting system

Internal linting is now promoted to a top-level tool alongside ESLint, Prettier, and tsc.

This fixes issues where skuba would not fail a `skuba lint` check but silently make changes to your working tree. These changes may have never been committed and subsequently led to noise when running `skuba format` or `skuba lint`.

Now, internal linting rules report whether changes need to be made, and changes are only applied in `format` or autofix modes (in CI).
```

**Review by 72636c — APPROVED** [review]
Bigly effort, ty 🙌

File: src/cli/lint/upgrade/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (72636c) [reply src/cli/lint/upgrade/index.ts]: 👍 Should we [rehash](https://github.com/seek-oss/skuba/blob/master/CHANGELOG.md#420) autofixes in the changeset for the stragglers? Could be something like:

> `skuba lint` can now automatically push autofixes. This eases adoption of linting rule changes and automatically resolves issues arising from a forgotten `skuba format`.
>
> You'll need to configure your CI environment to support this feature. See our [GitHub autofixes](https://seek-oss.github.io/skuba/docs/deep-dives/github.html#github-autofixes) documentation to learn more.

File: src/cli/configure/upgrade/patches/7.3.1/index.ts [hunk 7]
```diff
@@ -1,14 +1,26 @@
...
-    tryPatchRenovateConfig(),
-    tryPatchDockerfile(),
-    tryPatchServerListener(),
-  ]);
-};
+export const patches: Patches = [
+  {
+    apply: tryAddEmptyExports,
+    description:
+      'Add empty exports to Jest files for compliance with TypeScript isolated modules',
```
[the remark this answers is not in the document]
Comment (AaronMoat) [reply src/cli/configure/upgrade/patches/7.3.1/index.ts:12]: Yeah that could be good. Maybe if we give them IDs/names we could also later give a way to opt out of a change (e.g. node 20 later)

File: src/cli/lint/internalLints/refreshIgnoreFiles.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  };
+
+  const results = await Promise.all(
+    REFRESHABLE_IGNORE_FILES.map(refreshIgnoreFile),
+  );
+
+  // Log after for reproducible test output ordering
+  results.forEach((result) => {
+    if (result.msg) {
+      logger.warn(result.msg);
```
[the remark this answers is not in the document]
Comment (AaronMoat) [reply src/cli/lint/internalLints/refreshIgnoreFiles.ts:89]: :+1: sharing the name in a more centralised spot could be good, may tie into https://github.com/seek-oss/skuba/pull/1370#discussion_r1446747796

File: .changeset/five-dancers-pump.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+lint: Overhaul internal linting system
+
+Internal linting is now promoted to a top-level tool alongside ESLint, Prettier, and tsc.
+
+This fixes issues where skuba would not fail a `skuba lint` check but silently make changes to your working tree. These changes may have never been committed and subsequently led to noise when running `skuba format` or `skuba lint`.
+
+Now, internal linting rules report whether changes need to be made, and changes are only applied in `format` or autofix modes (in CI).
+
+In addition, `skuba lint` can now automatically push autofixes. This eases adoption of linting rule changes and automatically resolves issues arising from a forgotten `skuba format`.
```
Comment (AaronMoat) [inline .changeset/five-dancers-pump.md:13]: Should this be a separate changelog entry?

**Review by 72636c — APPROVED** [review]
🚀

**Comment by AaronMoat** [comment]
Thanks, you're very good at copyediting :)

**Pull request merged**